### PR TITLE
Device ID APIs

### PR DESCRIPTION
**Pull request by fredsadaghiani**
@mwylde PTAL
@sripadsriram FYI

### CODE REVIEW

File: HISTORY [hunk 1]
```diff
@@ -1,3 +1,7 @@
+=== 1.2.0.0 2015-04-12
```
Comment (mwylde) [inline HISTORY]: Date?

File: examples/device.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+TIMEOUT = 5 
+api_key = ENV['API_KEY']
+account_id = ENV['ACCOUNT_ID']
+
+
+if api_key.nil? || account_id.nil?
+  puts "Please specify the environment variables API_KEY and API_ID"
+  puts "usage:"
+  puts " API_KEY=xxxxxx API_ID=xxxxxxxx ruby device_id.rb"
```
Comment (mwylde) [inline examples/device.rb]: You use API_ID here and ACCOUNT_ID in line 11

File: examples/device.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  puts
+  exit
```
Comment (mwylde) [inline examples/device.rb]: Should probably have an error code — `exit 1`

File: examples/device.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+end
+
+class MyLogger
+  def warn(e)
+    puts "[WARN]  " + e.to_s
```
Comment (mwylde) [inline examples/device.rb]: these should probably all be to stderr — i.e., `$stderr.puts`

File: examples/device.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+def print_device(d)
+  puts 'i got the device obj: ' + d.inspect
+  puts 'device bad? ' + d.bad?.to_s
+  puts 'device labeled? ' + d.labeled?.to_s
+  puts 'device labeled_at ' + d.labeled_at.to_s
+  puts 'device session timestamps ' + d.session_timestamps.join(',')
+end
+
+device = Sift::Device.retrieve("i08a78q46cekukrugj7ltnqskj")
```
Comment (mwylde) [inline examples/device.rb]: Should this (and the session id) be specifiable on the command line?

More broadly, do we expect users to run this program or just use it as an example?

File: lib/sift/device.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    include Representable::JSON
+    include Representable::Object
+    property :time
+  end
+
+  class Timestamp < Struct.new :time
+  end
+
+  module ListRepresenter
+
```
Comment (mwylde) [inline lib/sift/device.rb]: nit: extra line

File: lib/sift/list.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+
+module Sift
```
Comment (mwylde) [inline lib/sift/list.rb]: What's this?

File: lib/sift/resource.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    module ClassMethods
+
+      BASE_URI = "https://api3.siftscience.com/v3/accounts/"
+
+      def resource_uri(path)
+        BASE_URI + Sift.account_id + path
+      end
+
+      def post(uri, options = {})
```
Comment (mwylde) [inline lib/sift/resource.rb]: Should this have an implementation (I realize it's not used now)?

File: lib/sift/resource.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      klass.extend(ClassMethods)
+      klass.include(InstanceMethods)
+    end
+
+    module ClassMethods
+
+      BASE_URI = "https://api3.siftscience.com/v3/accounts/"
+
+      def resource_uri(path)
+        BASE_URI + Sift.account_id + path
```
Comment (mwylde) [inline lib/sift/resource.rb]: Should we validate that Sift.account_id has been filled in?

File: lib/sift/resource.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+          },
+        }
+      end
+
+      def http_request(&blk)
+        response = blk.call
+        if response.success?
+          response.body
+        else
+          raise(RuntimeError, "Request failed: #{response.code}\n#{response.body}", caller)
```
Comment (mwylde) [inline lib/sift/resource.rb]: This isn't very friendly — raising a generic RuntimeException limits clients' ability to handle errors gracefully. There's also no easy way for them to get back the status code.

We may want to define our own exception hierarchy.

File: lib/sift/resource.rb [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        end
+      end
+
+    end
+
+    module InstanceMethods
```
Comment (mwylde) [inline lib/sift/resource.rb]: What's this?

**Comment by mwylde** [comment]
This broadly looks good to me, although we need to think a bit more about how errors should work.

File: HISTORY (hunk 1 above)
Comment (fredsadaghiani) [inline HISTORY]: done

File: examples/device.rb (hunk 2 above)
Comment (fredsadaghiani) [inline examples/device.rb]: done

File: examples/device.rb (hunk 3 above)
Comment (fredsadaghiani) [inline examples/device.rb]: done

File: examples/device.rb (hunk 4 above)
Comment (fredsadaghiani) [inline examples/device.rb]: done

File: examples/device.rb (hunk 5 above)
Comment (fredsadaghiani) [inline examples/device.rb]: It's simply meant to be an example program one could write. I'll update the fictitious id to be specified on the command line via an environment variable.

File: lib/sift/list.rb (hunk 7 above)
Comment (fredsadaghiani) [inline lib/sift/list.rb]: I'll remove it. I thought, at first, I could generalize list results. We can deal with this later.

File: lib/sift/resource.rb (hunk 9 above)
Comment (fredsadaghiani) [inline lib/sift/resource.rb]: Will do.

File: lib/sift/resource.rb (hunk 8 above)
Comment (fredsadaghiani) [inline lib/sift/resource.rb]: Added.

File: lib/sift/resource.rb (hunk 10 above)
Comment (fredsadaghiani) [inline lib/sift/resource.rb]: Done

File: lib/sift/resource.rb (hunk 11 above)
Comment (fredsadaghiani) [inline lib/sift/resource.rb]: Removed.

File: lib/sift/device.rb (hunk 6 above)
Comment (fredsadaghiani) [inline lib/sift/device.rb]: removed.

**Comment by fredsadaghiani** [comment]
@mwylde I've add a test and will add several more before asking for another review tomorrow.

**Comment by fredsadaghiani** [comment]
Otherwise, I've updated code to return native Ruby time objects (rather than timestamps in millis) throughout. See 2c010bf for details.

**Comment by yschatzberg** [comment]
@fredsadaghiani @mwylde What is the status of this PR?

**Pull request closed without merging**